OFEP-sdk-e2e-test-strategy.md

```diff
@@ -0,0 +1,61 @@
+## OFEP: SDK end-to-end test strategy
+
+## State: Drafting
+
+This OFEP proposes to introduce a simplified process to write SDK end-to-end(e2e) tests
+
+## Background
+
+SDKs are at the core of OpenFeature as they contain language specific implementations of the OpenFeature specification [1]
+
+For testing of the SDKs, current approach uses gherkin step definitions [2] contained in OpenFeature/test-harness [3] repository.
+And the test implementations use a real instance of flagd [4] to validate test rules.
+
+The current approach however comes with a problem. As SDK tests rely on flagd, these tests rely on SDK contributions [5].
+This means,
+
+- SDK cannot introduce tests till flagd contribution is there
+- Breaking changes require disabling e2e tests till
+  - Change propagate to respective SKD contribution repository
+  - flagd implements the breaking change
+- Cannot write tests till
+  - flagd implements the feature
+
+Besides, there is a circular dependency among `Go SDK` - `Go SDK Contribution` - `flagd` [6] which will become 
+harder to maintain in the long run.
+
+## Proposal
+
+Given OpenFeature has a lot of community interest and SDK, I am proposing to introduce a simplified e2e test strategy.
+The test strategy must comply with the following,
+
+- Tests must not use mocks and must run against an OpenFeature compliant provider
+- Tests must be self-contained except for test definitions which are common to all SDKs
+- Tests must verify error scenarios 
```

Comment:
```suggestion
Given OpenFeature has a lot of community interest and SDK, I am proposing to introduce a simplified e2e test strategy.
Gherkin tests must:
- not use mocks and must run against an OpenFeature compliant provider
- be self-contained except for test definitions which are common to all SDKs
- verify error scenarios 
```

I disagree with the wording of making this all tests, but if you're specifically referring to the gherkin tests, this seems fine with me.

Reply:
@justinabrahms Good point 🤔  my original intention of the OFEP was on e2e tests (Gherkin test suite). This is why I packaged the POC implementation of Go under `testing` [1]. And this closely follows the OTel testing framework  discussed here [2]

I am open to opinions here.

[1] - https://github.com/Kavindu-Dodan/go-sdk/tree/feat/in-memory-provider-for-testing/pkg/openfeature/testing
[2] - https://github.com/open-feature/ofep/pull/71#discussion_r1255278514

Reply:
I'm +1 w/ my proposed change.

Reply:
Just to confirm, your proposal is to make the in-memory provider SDK a built-in provider (not only for testing but as a fully-fledged provider). I can reword the proposal if there is no objection from others

Reply:
> Just to confirm, your proposal is to make the in-memory provider SDK a built-in provider (not only for testing but as a fully-fledged provider). I can reword the proposal if there is no objection from others

I'm not against this. Obviously it's real-work usage will have a limited value, but I think it's fine to consider it a "real" provider, with more usage beyond testing.

Reply:
Then I will reword the OFEP to highlight this fact. The in-memory provider will be bundled with SDK and not necessarily a test only provider.

Reply:
@justinabrahms & @toddbaert expanded the in-memory provider to be a general provider (not test only) [1]

[1] - https://github.com/open-feature/ofep/blob/ec965bcf29f7eab763f04a46c1cf19e1d55ba5f8/OFEP-sdk-e2e-test-strategy.md#in-memory-provider